internal/reconciler/metricpipeline/status.go

```diff
@@ -15,139 +16,97 @@ import (
 	}
 
-	return setCondition(ctx, r.Client, &pipeline, pending)
+	meta.SetStatusCondition(&pipeline.Status.Conditions, newCondition(conditions.TypeMetricAgentHealthy, reason, status, pipeline.Generation))
 }
 
-func (r *Reconciler) updateAgentStatus(ctx context.Context, pipelineName string) error {
-	log := logf.FromContext(ctx)
-
-	var pipeline telemetryv1alpha1.MetricPipeline
-	if err := r.Get(ctx, types.NamespacedName{Name: pipelineName}, &pipeline); err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil
-		}
-
-		return fmt.Errorf("failed to get MetricPipeline: %v", err)
+func (r *Reconciler) setGatewayHealthyCondition(ctx context.Context, pipeline *telemetryv1alpha1.MetricPipeline) {
+	gatewayName := types.NamespacedName{Name: r.config.Gateway.BaseName, Namespace: r.config.Gateway.Namespace}
+	healthy, err := r.gatewayProber.IsReady(ctx, gatewayName)
+	if err != nil {
+		logf.FromContext(ctx).V(1).Error(err, "Failed to probe metric gateway - set condition as not healthy")
+		healthy = false
 	}
 
-	agentReady, err := r.agentProber.IsReady(ctx, types.NamespacedName{Name: r.config.Agent.BaseName, Namespace: r.config.Agent.Namespace})
-	if err != nil {
-		return err
+	status := metav1.ConditionFalse
+	reason := conditions.ReasonMetricGatewayDeploymentNotReady
+	if healthy {
+		status = metav1.ConditionTrue
+		reason = conditions.ReasonMetricGatewayDeploymentReady
 	}
 
-	if agentReady {
-		if pipeline.Status.HasCondition(telemetryv1alpha1.MetricPipelineRunning) {
-			return nil
-		}
+	meta.SetStatusCondition(&pipeline.Status.Conditions, newCondition(conditions.TypeMetricGatewayHealthy, reason, status, pipeline.Generation))
+}
 
-		if pipeline.Status.HasCondition(telemetryv1alpha1.MetricPipelinePending) && pipeline.Status.Conditions[0].Reason == conditions.ReasonMetricGatewayDeploymentNotReady {
-			return nil
-		}
-		running := telemetryv1alpha1.NewMetricPipelineCondition(conditions.ReasonMetricAgentDaemonSetReady, telemetryv1alpha1.MetricPipelineRunning)
-		return setCondition(ctx, r.Client, &pipeline, running)
+func (r *Reconciler) setGatewayConfigGeneratedCondition(ctx context.Context, pipeline *telemetryv1alpha1.MetricPipeline, lockAcquired bool) {
+	status := metav1.ConditionTrue
+	reason := conditions.ReasonMetricConfigurationGenerated
+
+	if !lockAcquired {
+		status = metav1.ConditionFalse
+		reason = conditions.ReasonWaitingForLock
 	}
 
-	pendingAgent := telemetryv1alpha1.NewMetricPipelineCondition(conditions.ReasonMetricAgentDaemonSetNotReady, telemetryv1alpha1.MetricPipelinePending)
-	if pipeline.Status.HasCondition(telemetryv1alpha1.MetricPipelineRunning) {
-		log.V(1).Info(fmt.Sprintf("Updating the status of %s to %s. Resetting previous conditions", pipeline.Name, pendingAgent.Type))
-		pipeline.Status.Conditions = []telemetryv1alpha1.MetricPipelineCondition{}
+	referencesNonExistentSecret := secretref.ReferencesNonExistentSecret(ctx, r.Client, pipeline)
```

Comment:
I'm wondering if we should change the order here. My first intuition is that the pipeline limit is the "hardest" condition and if that is reached, we should indicate that we entirely ignore the pipeline.

Reply:
Yeah, I agree. I will change it and adapt the tests